src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -520,13 +520,21 @@ void GcsActorManager::DestroyActor(const ActorID &actor_id) {
       // The actor was being scheduled and has now been canceled.
       RAY_CHECK(canceled_actor_id == actor_id);
     } else {
-      // The actor was pending scheduling. Remove it from the queue.
       auto pending_it = std::find_if(pending_actors_.begin(), pending_actors_.end(),
                                      [actor_id](const std::shared_ptr<GcsActor> &actor) {
                                        return actor->GetActorID() == actor_id;
                                      });
-      RAY_CHECK(pending_it != pending_actors_.end());
-      pending_actors_.erase(pending_it);
+      // The actor was pending scheduling. Remove it from the queue.
+      if (pending_it != pending_actors_.end()) {
+        pending_actors_.erase(pending_it);
+      } else {
+        // When actor creation request of this actor id is pending in raylet, we should
+        // cancel lease. NOTE(sang): It is possible the raylet replies to the lease
+        // request after this function is called. It will be fine because at the end of
+        // this function, we will publish the dead actor state, and it will make the
+        // raylet return the worker that was leased.
+        gcs_actor_scheduler_->CancelLeasingRequest(node_id, actor_id);
```

Comment:
If the result of the request worker lease is returned after this line is executed, we should send a message to the raylet to cancel the lease worker?

Reply:
At the end of this method, we'll publish the raylet that this actor is dead. Raylet handles this scenario at `HandleActorStateTransition`. 

One concerning I found after your comment is `HandleActorStateTransition`'s DEAD actor code and `HandleWorkerLeaseRequest`'s code is a little bit different. 

@stephanie-wang, are they working in the same way?

Reply:
I went deep down to the code, and it seems like `HandleActorStateTransition ` is for actor and `CancelLeaseWorker` is for tasks. Dead actor workers lease should be returned by `HandleActorStateTransition`. That means this should handle the case. I'll wait until @stephanie-wang confirms it.

Reply:
Hmm I'm not sure if the node manager actually does that when receiving the notification. How about we do something similar to normal tasks where we just return the lease immediately if the actor has already been destroyed? It'd be good to centralize as much of the logic in the GCS as possible.

Reply:
Got it. I will then make it call the cancelWorkerLease request here!

Reply:
Actually, let me do some experiment before we do this. I will get back to you.

Reply:
Just confirm, it correctly returns the lease request without explicitly calling `CancelLeaseRequest`. I think they are cleaned by published actor dead state messages. 

I will probably just add a TODO message here just in case. I will also add one more integration test to verify this scenario.

Reply:
Added TODO instead. I understand explicitly cancelling lease to raylet is a better implementation, so please let me know if you think that's the blocker for this PR.

Reply:
Hmm I think it does cancel the lease request if the owner dies, so I think you're right that we can remove the RPC to the raylet (and the TODOs).

Reply:
Got it. I will remove the todo here! One concern I have is that people might misunderstand this method will actually cancel lease request. I will explicitly mention this in the method declaration instead!

Reply:
Yeah maybe also a different name for the method.

Reply:
Decided to use `CancelOnLeasing`, which is consistent to other methods (`CancelOnWorker` or `CacnelOnNode`).